source/extensions/clusters/redis/redis_cluster_lb.h

```diff
@@ -37,7 +37,18 @@ class ClusterSlot {
   int64_t start() const { return start_; }
   int64_t end() const { return end_; }
   Network::Address::InstanceConstSharedPtr primary() const { return primary_; }
-  const absl::flat_hash_set<Network::Address::InstanceConstSharedPtr>& replicas() const {
+  struct Hash {
+    size_t operator()(const Network::Address::InstanceConstSharedPtr& address) const {
+      return absl::Hash<std::string>()(address->asString());
```

Comment:
As a matter of good practice, you'd want to use `return absl::Hash<absl::string_view>()(address->asStringView());` to avoid potentially creating a string temp while hashing. Though actually the name is held in a string in the main impl, so the `asString()` impl currently won't do that.

But in researching this I found a potential latent crash here. The doc for `asString()` says:
```
   * @return a human readable string for the address that represents the
   * physical/resolved address. (This will not necessarily include port
   * information, if applicable, since that may not be resolved until bind()).
```

This suggests to me that bind() might mutate the name. If that happens after we have already stored an address in a hash-table, I think this can crash in subsequent hash-table operations.

So my suggestion is to convert this to a `map<string,Network::Address::InstanceConstSharedPtr>`. This will result in string-copies for the map key, but will avoid crashing Envoy if `bind()` mutates the name. And if you do that you won't need this custom hasher/comparator.

Reply:
+1

That's a good catch! There's no good in hashing keys which can mutate.

Reply:
Thanks for this suggestion, I have use map instead of flat_hash_set and create a commit.